### PR TITLE
frontend: use zero-allocation response decoder in shard active series middleware 

### PR DESCRIPTION
**Pull request by ortuman**
#### What this PR does

At certain times we've observed in ops huge spikes of allocations during the deserialization/processing of each of the query responses when merging the results in the shard active series frontend middleware, occasionally reaching values close to `~80%` of total allocations.

Screenshot 2024-03-21 at 09 28 42

While this condition does not critically compromise the service, nor does imply a customer impact, such spikes translate into high CPU consumption that can negatively impact the TCO.

This PR aims to solve this problem through the implementation of a zero-allocation decoder, specifically adapted to the use case of said middleware, that borrows certain functions from the json-iter library, currently in use.

Broadly speaking, the way this new decoder works is by copying all contents of the response `data` key into an internal buffer that is subsequently reused once is received and consumed by the merger goroutine, rather than copying each of the strings associated with all label-value pairs.

_Below are the comparative benchmark results before and after the change._

```
name                                                                         old time/op    new time/op    delta
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-4-8           8.93µs ± 0%    7.51µs ± 3%   -15.94%  (p=0.016 n=4+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-16-8          15.9µs ± 1%    15.9µs ± 2%      ~     (p=0.690 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-64-8          45.0µs ± 1%    43.3µs ± 2%    -3.75%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-128-8         77.8µs ± 0%    79.2µs ± 3%      ~     (p=0.222 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-4-8         14.5µs ± 1%    12.7µs ± 1%   -12.09%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-16-8        25.4µs ± 3%    24.0µs ± 1%    -5.66%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-64-8        59.9µs ± 1%    61.2µs ± 0%    +2.22%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-128-8       91.7µs ± 2%    98.7µs ± 2%    +7.58%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-4-8       1.90ms ± 0%    0.51ms ± 0%   -72.95%  (p=0.016 n=5+4)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-16-8      7.22ms ± 0%    2.10ms ± 8%   -70.95%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-64-8      28.3ms ± 1%     7.7ms ±11%   -72.63%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-128-8     56.1ms ± 1%    16.0ms ±12%   -71.52%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-4-8      18.8ms ± 0%     5.0ms ± 0%   -73.26%  (p=0.016 n=5+4)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-16-8     72.2ms ± 0%    22.6ms ±11%   -68.65%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-64-8      281ms ± 1%      81ms ± 4%   -71.23%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-128-8     558ms ± 3%     150ms ± 1%   -73.13%  (p=0.008 n=5+5)

name                                                                         old alloc/op   new alloc/op   delta
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-4-8           1.98kB ± 0%    1.80kB ± 2%    -9.07%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-16-8          3.90kB ± 0%    2.91kB ± 1%   -25.31%  (p=0.016 n=4+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-64-8          11.6kB ± 0%     8.7kB ± 1%   -24.79%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-128-8         22.0kB ± 0%    19.8kB ± 1%    -9.67%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-4-8         2.75kB ± 5%    2.55kB ± 5%    -7.41%  (p=0.016 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-16-8        4.91kB ± 6%    3.92kB ± 6%   -20.26%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-64-8        13.0kB ± 4%    11.2kB ± 4%   -14.07%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-128-8       23.5kB ± 4%    24.3kB ±10%      ~     (p=0.222 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-4-8        406kB ± 0%      95kB ± 0%   -76.58%  (p=0.016 n=5+4)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-16-8      1.37MB ± 0%    0.12MB ± 2%   -91.19%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-64-8      5.25MB ± 0%    0.43MB ±24%   -91.80%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-128-8     10.5MB ± 0%     1.2MB ±59%   -88.51%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-4-8      3.29MB ± 0%    0.88MB ± 0%   -73.38%  (p=0.016 n=5+4)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-16-8     12.9MB ± 0%     2.0MB ± 3%   -84.61%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-64-8     51.7MB ± 0%     5.4MB ±12%   -89.61%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-128-8     104MB ± 0%      17MB ±14%   -83.32%  (p=0.008 n=5+5)

name                                                                         old allocs/op  new allocs/op  delta
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-4-8             60.0 ± 0%      25.0 ± 0%   -58.33%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-16-8             192 ± 0%        49 ± 0%   -74.48%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-64-8             720 ± 0%       145 ± 0%   -79.86%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-128-8          1.42k ± 0%     0.27k ± 0%   -80.83%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-4-8           67.0 ± 0%      32.0 ± 0%   -52.24%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-16-8           199 ± 0%        56 ± 0%   -71.86%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-64-8           727 ± 0%       152 ± 0%   -79.09%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/encoding=snappy/num-responses-128-8        1.43k ± 0%     0.28k ± 0%   -80.43%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-4-8        32.1k ± 0%      0.0k ± 0%   -99.91%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-16-8        128k ± 0%        0k ± 0%      ~     (p=0.079 n=4+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-64-8        513k ± 0%        0k ± 0%   -99.97%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=1_000/num-responses-128-8      1.03M ± 0%     0.00M ± 1%   -99.97%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-4-8        320k ± 0%        0k ± 0%      ~     (p=0.079 n=4+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-16-8      1.28M ± 0%     0.00M ± 0%  -100.00%  (p=0.000 n=5+4)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-64-8      5.12M ± 0%     0.00M ± 2%  -100.00%  (p=0.008 n=5+5)
ActiveSeriesMiddlewareMergeResponses/seriesCount=10_000/num-responses-128-8     10.2M ± 0%      0.0M ± 4%  -100.00%  (p=0.008 n=5+5)
```

_Raw benchmark results_:
* [old.bench](https://gist.github.com/ortuman/abfb50aace899ca08bf437cd0689ad12)  
* [new.bench](https://gist.github.com/ortuman/912707436a5c828d95cb493e4285d952)

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir-squad/issues/1870

#### Checklist

- [X] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by flxbk** [comment]
Cool, thanks for working on this! I'll try to give it a closer look soon, but one thing that occurred to me: Do you think it's worth and feasible putting this behind a feature flag to make rollback easier in case any unexpected issues appear? Not sure how risky you deem this change.

**Comment by ortuman** [comment]
> Do you think it's worth and feasible putting this behind a feature flag to make rollback easier in case any unexpected issues appear?

I believe it makes complete sense and is perfectly viable. Once we've tested that everything works as expected we can get rid of the old code. 👍 

I'll see to rework the PR to allow both code paths. In the meantime, I'll mark the PR as a draft again.

**Comment by ortuman** [comment]
Added new `-query-frontend.use-shard-active-series-zero-allocation-decoder` experimental option to enable/disable this feature at will. 

_@flxbk @Logiraptor this is ready for review now_

File: cmd/mimir/config-descriptor.json [hunk 1]
```diff
@@ -5682,6 +5682,17 @@
           "fieldType": "boolean",
           "fieldCategory": "experimental"
         },
+        {
+          "kind": "field",
+          "name": "use_shard_active_series_zero_allocation_decoder",
+          "required": false,
+          "desc": "True to use the zero-allocation response decoder for active series queries.",
```
Comment (jdbaldry) [inline cmd/mimir/config-descriptor.json:5689]:
```suggestion
          "desc": "Set to true to use the zero-allocation response decoder for active series queries.",
```

File: pkg/frontend/querymiddleware/roundtrip.go [hunk 2]
```diff
@@ -81,6 +82,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.Uint64Var(&cfg.TargetSeriesPerShard, "query-frontend.query-sharding-target-series-per-shard", 0, "How many series a single sharded partial query should load at most. This is not a strict requirement guaranteed to be honoured by query sharding, but a hint given to the query sharding when the query execution is initially planned. 0 to disable cardinality-based hints.")
 	f.StringVar(&cfg.QueryResultResponseFormat, "query-frontend.query-result-response-format", formatProtobuf, fmt.Sprintf("Format to use when retrieving query results from queriers. Supported values: %s", strings.Join(allFormats, ", ")))
 	f.BoolVar(&cfg.ShardActiveSeriesQueries, "query-frontend.shard-active-series-queries", false, "True to enable sharding of active series queries.")
+	f.BoolVar(&cfg.UseShardActiveSeriesZeroAllocDecoder, "query-frontend.use-shard-active-series-zero-allocation-decoder", false, "True to use the zero-allocation response decoder for active series queries.")
```
Comment (flxbk) [inline pkg/frontend/querymiddleware/roundtrip.go:85]: This is a mouthful, wdyt about `-query-frontend.use-active-series-decoder`?

File: pkg/frontend/querymiddleware/shard_active_series_response_decoder.go [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+				inner--
+				if inner == 0 {
+					d.foundDataValue = true
+					return
+				}
+			default:
+				d.out = append(d.out, c)
+			}
+			c = d.readByte()
+		}
```
Comment (flxbk) [inline pkg/frontend/querymiddleware/shard_active_series_response_decoder.go:162]: If I understand correctly this reads the entire value contained in this response's data field into `d.out`? If yes I think we need to revisit this, because this would lead to the same kind of OOMs that we had before response streaming if I'm not mistaken 🤔

File: pkg/frontend/querymiddleware/shard_active_series_test.go [hunk 4]
```diff
@@ -423,7 +425,7 @@ func Test_shardActiveSeriesMiddleware_RoundTrip_concurrent(t *testing.T) {
 }
 
 func Test_shardActiveSeriesMiddleware_mergeResponse_contextCancellation(t *testing.T) {
-	s := newShardActiveSeriesMiddleware(nil, mockLimits{}, log.NewNopLogger()).(*shardActiveSeriesMiddleware)
+	s := newShardActiveSeriesMiddleware(nil, true, mockLimits{}, log.NewNopLogger()).(*shardActiveSeriesMiddleware)
```
Comment (flxbk) [inline pkg/frontend/querymiddleware/shard_active_series_test.go:428]: As long as both code paths are still there I think it makes sense to run this test case for both, wdyt?

File: pkg/frontend/querymiddleware/shard_active_series_test.go [hunk 5]
```diff
@@ -440,19 +442,19 @@ func Test_shardActiveSeriesMiddleware_mergeResponse_contextCancellation(t *testi
...
+
 	resp := s.mergeResponses(ctx, responses, "")
 	defer func() {
 		_, _ = io.Copy(io.Discard, resp.Body)
 		_ = resp.Body.Close()
 	}()
 
 	var buf bytes.Buffer
-	_, err = io.CopyN(&buf, resp.Body, int64(os.Getpagesize()))
+	_, err = io.Copy(&buf, resp.Body)
```
Comment (flxbk) [inline pkg/frontend/querymiddleware/shard_active_series_test.go:455]: Why are these changes needed?

**Review by flxbk — COMMENTED** [review]
I had a few questions, mainly about what we might be accumulating into memory

File: pkg/frontend/querymiddleware/shard_active_series_response_decoder.go (hunk 3 above)
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series_response_decoder.go:162]: Sure, during implementation I didn't realize that these data fields could end up being several megabytes in length. 🤦‍♂️  

In that case, I think it would make all the sense to  send `data` array content through a channel instead, to avoid the use of an intermediate buffer. wdyt?

File: pkg/frontend/querymiddleware/shard_active_series_test.go (hunk 4 above)
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series_test.go:428]: sounds fair! 👍

File: pkg/frontend/querymiddleware/shard_active_series_test.go (hunk 5 above)
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series_test.go:455]: I believe it was due to a side effect caused by the fact that now the data value is preprocessed before performing the merge operation, but since this logic is going to change due to https://github.com/grafana/mimir/pull/7665#discussion_r1535163838, I'll restore the old code.

File: pkg/frontend/querymiddleware/shard_active_series_response_decoder.go (hunk 3 above)
Comment (flxbk) [reply pkg/frontend/querymiddleware/shard_active_series_response_decoder.go:162]: Yes, the `data` field of the responses can be several hundred MBs large. The other implementation sends each item of the `data` array separately once it's parsed (see [here](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/shard_active_series.go#L335)), so I think it makes sense to do something equivalent in this approach as well.

**Comment by ortuman** [comment]
_pinging @flxbk_

ready for review again 👍

**Review by flxbk — COMMENTED** [review]
I had a question about how we're using channels to forward bits of responses to the merging code.

File: pkg/frontend/querymiddleware/shard_active_series.go [hunk 6]
```diff
@@ -434,3 +486,89 @@ func (s *shardActiveSeriesMiddleware) writeMergedResponse(ctx context.Context, c
...
+		}
+
+		for {
+			chunkStr := unsafe.String(
+				unsafe.SliceData(chunk.buff.Bytes()),
+				chunk.buff.Len(),
+			)
+
+			// Write the value as is, since it's already a JSON array.
+			stream.WriteRaw(chunkStr)
```
Comment (Logiraptor) [inline pkg/frontend/querymiddleware/shard_active_series.go:544]: Any reason not to use `Write()` here and skip the unsafe string conversion?

**Review by Logiraptor — COMMENTED** [review]
As far as I understand, the main thing in the PR that causes zero allocations is that it assumes the responses coming from queriers are valid json, and then proceeds to shuffle around slices of bytes based on `[` and `]` as delimiters. Neat!

I do worry about debugging this if it ever breaks, so I added a fuzz test which compares this PR to the stdlib json result. The first failure is the input `"{\"data\":[]"` due to the missing `}`. The code in the PR accepts that invalid json. I understand that this code is extremely fast, but it makes me nervous at the same time. Do you think making the code fail on invalid json will eliminate the performance gains?

Here's the fuzz test as an example:

```go
func FuzzActiveSeriesResponseDecoder(f *testing.F) {
	inputs := []string{
		`{"data":[]}`,
		`{"unexpected_1":3.141516, "unexpected_2":"skip me", "unexpected_3":[[{}]], "unexpected_4": {"key":[]}, "unexpected_5":null, "unexpected_6":true, "data":[{"__name__":"metric","shard":"1"}]}`,
		`{"data":[{"__name__":"metric","shard":"1"},{"__name__":"metric","shard":"2"}]}`,
	}

	for _, input := range inputs {
		f.Add(input)
	}

	f.Fuzz(func(t *testing.T, input string) {
		var dataStr strings.Builder

		errCh := make(chan error, 1)

		r := strings.NewReader(input)
		d := borrowShardActiveSeriesResponseDecoder(context.Background(), io.NopCloser(r))
		chunkCh, err := d.decode()
		if err == nil {
			go func() {
				errCh <- d.streamData()
			}()

			// Drain the data channel.
			for chunk := range chunkCh {
				if chunk.done {
					break
				}
				dataStr.WriteString(chunk.buff.String())
			}
		} else {
			errCh <- err
		}

		err = <-errCh

		type response struct {
			Data []map[string]string `json:"data"`
		}

		var fizzInputResult response
		errOnFuzzInput := json.NewDecoder(strings.NewReader(input)).Decode(&fizzInputResult)
		var decoderOutputResult interface{}
		errOnDecoderOutput := json.NewDecoder(strings.NewReader(dataStr.String())).Decode(&decoderOutputResult)

		if errOnFuzzInput != nil {
			require.Error(t, err)
		} else {
			require.NoError(t, err)
			require.NoError(t, errOnDecoderOutput)
			require.Equal(t, fizzInputResult, decoderOutputResult)
		}
	})
}
```

And I ran it like so:

```
go test -timeout 30s -fuzz ^FuzzActiveSeriesResponseDecoder$ -run=xxx ./pkg/frontend/querymiddleware
```

File: pkg/frontend/querymiddleware/shard_active_series.go [hunk 7]
```diff
@@ -434,3 +486,89 @@ func (s *shardActiveSeriesMiddleware) writeMergedResponse(ctx context.Context, c
...
+
+	stream.WriteObjectStart()
+	stream.WriteObjectField("data")
+	stream.WriteArrayStart()
+	firstItem := true
+	for chunkCh := range items {
+		chunk := <-chunkCh
+		if chunk.done {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series.go]: _addressed @ 203b8ed6234f9db80fbf96933c21dc91193a2324_

File: pkg/frontend/querymiddleware/shard_active_series_response_decoder.go [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+		return bytes.NewBuffer(make([]byte, 0, activeSeriesChunkMaxBufferSize))
+	},
+}
+
+var shardActiveSeriesResponseDecoderPool = sync.Pool{
+	New: func() any {
+		return &shardActiveSeriesResponseDecoder{
+			br:      bufio.NewReaderSize(nil, 4096),
+			strBuff: make([]byte, 0, 256),
+			chunkCh: make(chan activeSeriesDataChunk),
```
[the remark this answers is not in the document]
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series_response_decoder.go]: _addressed @ 203b8ed6234f9db80fbf96933c21dc91193a2324_

File: CHANGELOG.md [hunk 9]
```diff
@@ -11,6 +11,7 @@
 * [ENHANCEMENT] Store-gateway: merge series from different blocks concurrently. #7456
 * [ENHANCEMENT] Store-gateway: Add `stage="wait_max_concurrent"` to `cortex_bucket_store_series_request_stage_duration_seconds` which records how long the query had to wait for its turn for `-blocks-storage.bucket-store.max-concurrent`. #7609
 * [ENHANCEMENT] Querier: add `cortex_querier_federation_upstream_query_wait_duration_seconds` to observe time from when a querier picks up a cross-tenant query to when work begins on its single-tenant counterparts. #7209
+* [ENHANCEMENT] Query-frontend: use zero-allocation experimental decoder for active series queries via `-query-frontend.use-shard-active-series-zero-allocation-decoder`. #7665
```
[the remark this answers is not in the document]
Comment (ortuman) [reply CHANGELOG.md]: _addressed @ f8268e988ba63ad364d7385a946e7f6de15d0586_

File: pkg/frontend/querymiddleware/shard_active_series.go (hunk 6 above)
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series.go]: main reason is that by using of `*stream.Write` method, which also invokes an underlying `io.Writer` interface, forces the compiler to escape the `buf.Bytes()` slice to the heap, due to the fact that the compiler has no way of knowing whether or not the slice is retained by the implementation. anyway, this does not make a huge difference in total number of allocations, so if you consider it's worth making the sacrifice, we can make the change.

File: pkg/frontend/querymiddleware/shard_active_series_response_decoder.go [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+				return d.err
+			}
+		}
+		cb.WriteByte(c)
+
+		if cb.Len() >= activeSeriesChunkMaxBufferSize {
+			d.chunkCh <- activeSeriesDataChunk{buff: cb}
+			cb = activeSeriesChunkBufferPool.Get().(*bytes.Buffer)
+			d.checkContextCanceled()
+		}
```
[the remark this answers is not in the document]
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series_response_decoder.go]: _changed @ f9e62dd950339dcc73b0a1714e3ef1f25736a16a_

File: pkg/frontend/querymiddleware/shard_active_series.go [hunk 11]
```diff
@@ -356,6 +363,51 @@ func (s *shardActiveSeriesMiddleware) mergeResponses(ctx context.Context, respon
 	return resp
 }
 
+func (s *shardActiveSeriesMiddleware) mergeResponsesWithZeroAllocationDecoder(ctx context.Context, responses []*http.Response, encoding string) *http.Response {
+	reader, writer := io.Pipe()
+
+	items := make(chan chan activeSeriesDataChunk, len(responses))
```
[the remark this answers is not in the document]
Comment (ortuman) [reply pkg/frontend/querymiddleware/shard_active_series.go]: _changed @ f9e62dd950339dcc73b0a1714e3ef1f25736a16a_

**Comment by ortuman** [comment]
> I do worry about debugging this if it ever breaks, so I added a fuzz test which compares this PR to the stdlib json result. The first failure is the input "{\"data\":[]" due to the missing }. The code in the PR accepts that invalid json. I understand that this code is extremely fast, but it makes me nervous at the same time. Do you think making the code fail on invalid json will eliminate the performance gains?

If I remember correctly, the current implementation would also accept the input `{"data": [...]`, as the processing of the querier response stops as soon as the content of the `data` key has been processed.

Also, after the change in 23e736b4a289ba1efaa27651d984402c3e6e1fd2, at least now we ensure that the content of said key at least takes the form `[{...},{...}, ...]`, although it is true that the objects contained in the array can contain any random data, and such content would be propagated to the frontend. Which, on the other hand, I don't see as a cause for concern. If the querier returns corrupt data, it would make sense for the frontend to propagate such data, thereby highlighting the problem instead of covering it up.

**Comment by ortuman** [comment]
_pinging @flxbk @Logiraptor for a new review_

**Review by Logiraptor — APPROVED** [review]

**Review by flxbk — APPROVED** [review]
LGTM
